xiaogpt/tts/volc.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+import logging
+import tempfile
+from pathlib import Path
+from typing import Optional
+import json
+import os
+import time
+import base64
+import threading
+
+from volcengine.ApiInfo import ApiInfo
+from volcengine.Credentials import Credentials
+from volcengine.ServiceInfo import ServiceInfo
+from volcengine.base.Service import Service
+
+
+from xiaogpt.tts.base import AudioFileTTS
+from xiaogpt.utils import calculate_tts_elapse
+import requests
+
+logger = logging.getLogger(__name__)
+
+
+class VolcTTS(AudioFileTTS):
+    def __init__(self, mina_service, device_id, config):
+        super().__init__(mina_service, device_id, config)
+        self.token = get_token(config)
+        logger.info("Initializing VolcTTS {self.token}")
+    
+    async def make_audio_file(self, query: str, text: str) -> tuple[Path, float]:
+        tts_payload = json.dumps({
+            "text": text,
+            "speaker": self.config.volc_tts_speaker,
+            "audio_config": {
+                "format": "mp3",
+                "sample_rate": 24000,
+                "speech_rate": 0,
+            },
+        })
+        
+        
+        req = {
+            "appkey": self.config.volc_tts_app,
+            "token": self.token,
+            "namespace": "TTS",
+            "payload": tts_payload,
+        }
+        
+        resp = requests.post("https://sami.bytedance.com/api/v1/invoke", json=req)
+        try:
+            sami_resp = resp.json()
+            logger.info(f"sami_resp {resp.status_code}")
+            if resp.status_code != 200:
+                print(sami_resp)
+        except:
+            logger.error(f"Failed to get tts from volcengine with voice=zh {text}")
+        
+        if sami_resp["status_code"] == 20000000 and len(sami_resp["data"]) > 0:
+            audio_data = base64.b64decode(sami_resp["data"])
+            with tempfile.NamedTemporaryFile(
+                suffix=".mp3", mode="wb", delete=False, dir=self.dirname.name
+            ) as f:
+                f.write(audio_data)
+                
+        return Path(f.name), calculate_tts_elapse(text)
+
+## fetch token and save it to file
+def get_token(config):
+    token_file = "/tmp/volc_token.json"
+    if not os.path.exists(token_file):
+        token = request_token_data(config)
+    else:
+        with open(token_file, "r") as f:
+            token = json.load(f)
+            if token['expires_at'] < time.time():
+                token = request_token_data(config)
```

Comment:
this path is not support windows I wonder

Reply:
how about making it to be `token_file = "/tmp/volc_token.json" if os.path.exists("/tmp/") else os.path.join(os.getenv('TEMP'), 'volc_token.json')` ?

If we don't save it into the file, it may throw errors if we restart the server (ioken is requested too frequently)

Reply:
you can use pathlib for both win and linux https://docs.python.org/3/library/pathlib.html